hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -1097,87 +1165,76 @@ protected void cleanIfNecessary(BaseHoodieWriteClient writeClient, String instan
-            engineContext, Collections.emptyMap(), partitionToFilesMap, getRecordsGenerationParams(), createInstantTime);
-        partitionToRecordsMap.put(MetadataPartitionType.BLOOM_FILTERS, recordsRDD);
-      }
-
-      if (partitionTypes.contains(MetadataPartitionType.COLUMN_STATS) && totalDataFilesCount > 0) {
-        final HoodieData<HoodieRecord> recordsRDD = HoodieTableMetadataUtil.convertFilesToColumnStatsRecords(
-            engineContext, Collections.emptyMap(), partitionToFilesMap, getRecordsGenerationParams());
-        partitionToRecordsMap.put(MetadataPartitionType.COLUMN_STATS, recordsRDD);
-      }
-      LOG.info("Committing " + partitions.size() + " partitions and " + totalDataFilesCount + " files to metadata");
+  private boolean validateTimelineBeforeSchedulingCompaction(Option<String> inFlightInstantTimestamp, String latestDeltacommitTime) {
+    // There should not be any incomplete instants on MDT
+    HoodieActiveTimeline metadataTimeline = metadataMetaClient.reloadActiveTimeline();
+    List<HoodieInstant> pendingInstantsOnMetadataTable = metadataTimeline.filterInflightsAndRequested().getInstants();
+    if (!pendingInstantsOnMetadataTable.isEmpty()) {
+      LOG.info(String.format(
+              "Cannot compact MDT as there are %d inflight instants: %s",
+              pendingInstantsOnMetadataTable.size(), Arrays.toString(pendingInstantsOnMetadataTable.toArray())));
+      metrics.ifPresent(m -> m.incrementMetric(HoodieMetadataMetrics.SKIP_TABLE_SERVICES, 1));
+      return false;
     }
 
-    commit(createInstantTime, partitionToRecordsMap, false);
-  }
+    // There should not be any incomplete instants on dataset
+    HoodieActiveTimeline datasetTimeline = dataMetaClient.reloadActiveTimeline();
+    List<HoodieInstant> pendingInstantsOnDataset = datasetTimeline.filterInflightsAndRequested().getInstantsAsStream()
+            .filter(i -> !inFlightInstantTimestamp.isPresent() || !i.getTimestamp().equals(inFlightInstantTimestamp.get()))
+            .collect(Collectors.toList());
+    if (!pendingInstantsOnDataset.isEmpty()) {
+      LOG.info(String.format(
+              "Cannot compact MDT as there are %d inflight instants on dataset before latest deltacommit %s: %s",
+              pendingInstantsOnDataset.size(), latestDeltacommitTime, Arrays.toString(pendingInstantsOnDataset.toArray())));
+      metrics.ifPresent(m -> m.incrementMetric(HoodieMetadataMetrics.SKIP_TABLE_SERVICES, 1));
+      return false;
+    }
 
-  private HoodieData<HoodieRecord> getFilesPartitionRecords(String createInstantTime, List<DirectoryInfo> partitionInfoList, HoodieRecord allPartitionRecord) {
-    HoodieData<HoodieRecord> filesPartitionRecords = engineContext.parallelize(Arrays.asList(allPartitionRecord), 1);
-    if (partitionInfoList.isEmpty()) {
-      return filesPartitionRecords;
+    // Check if the inflight commit is greater than all the completed commits.
```

Comment:
do we really need to add this constraint? again, can you check if L1215 - 1234 would suffice. I am inclining to favor the liveness of MDT compaction. if these are added just to be very cautious and we don't really need it, we should remove it.

Reply:
I have simplified it and removed the extra checks.